File: frontend/src/api/installationRequests.ts
import { api } from "./base";

export interface InstallationItem {
  id: string;
  room_id?: string;
  coverage_type: string;
  desired_device_count: number;
  notes?: string;
  proposed_device_type?: string;
  status: string;
}

export interface InstallationRequest {
  id: string;
  home_id: string;
  owner_id: string;
  technician_id?: string;
  status: string;
  notes?: string;
  created_at: string;
  updated_at: string;
  items: InstallationItem[];
}

export interface InstallationItemCreate {
  room_id?: string;
  coverage_type: string;
  desired_device_count?: number;
  notes?: string;
}

export interface InstallationRequestCreate {
  home_id?: string;
  notes?: string;
  items: InstallationItemCreate[];
}

export interface OwnerInstallationStatusUpdate {
  action: "approve" | "request_changes";
  notes?: string;
}

export interface TechInstallationUpdate {
  status?: "in_review" | "plan_ready" | "installed";
  notes?: string;
}

export interface TechInstallationItemUpdate {
  status?: "pending" | "approved" | "rejected" | "installed";
  desired_device_count?: number;
  notes?: string;
}

export const installationRequestsApi = api.injectEndpoints({
  endpoints: (builder) => ({
    listOwnerInstallationRequests: builder.query<InstallationRequest[], void>({
      query: () => ({
        url: "/owner/installation-requests",
        method: "GET",
      }),
      providesTags: ["InstallationRequest"],
    }),
    createOwnerInstallationRequest: builder.mutation<
      InstallationRequest,
      InstallationRequestCreate
    >({
      query: (body) => ({
        url: "/owner/installation-requests",
        method: "POST",
        body,
      }),
      invalidatesTags: ["InstallationRequest"],
    }),
    updateOwnerInstallationRequest: builder.mutation<
      InstallationRequest,
      { id: string; data: OwnerInstallationStatusUpdate }
    >({
      query: ({ id, data }) => ({
        url: `/owner/installation-requests/${id}`,
        method: "PATCH",
        body: data,
      }),
      invalidatesTags: ["InstallationRequest"],
    }),
    listTechInstallationRequests: builder.query<
      InstallationRequest[],
      { status?: string } | void
    >({
      query: (params) => ({
        url: "/tech/installation-requests",
        method: "GET",
        params: params ? params : undefined,
      }),
      providesTags: ["InstallationRequest"],
    }),
    updateTechInstallationRequest: builder.mutation<
      InstallationRequest,
      { id: string; data: TechInstallationUpdate }
    >({
      query: ({ id, data }) => ({
        url: `/tech/installation-requests/${id}`,
        method: "PATCH",
        body: data,
      }),
      invalidatesTags: ["InstallationRequest"],
    }),
    updateTechInstallationItem: builder.mutation<
      InstallationRequest,
      { requestId: string; itemId: string; data: TechInstallationItemUpdate }
    >({
      query: ({ requestId, itemId, data }) => ({
        url: `/tech/installation-requests/${requestId}/items/${itemId}`,
        method: "PATCH",
        body: data,
      }),
      invalidatesTags: ["InstallationRequest"],
    }),
    approveAllTechInstallationItems: builder.mutation<
      InstallationRequest,
      { requestId: string }
    >({
      query: ({ requestId }) => ({
        url: `/tech/installation-requests/${requestId}/approve-all`,
        method: "POST",
      }),
      invalidatesTags: ["InstallationRequest"],
    }),
  }),
});

export const {
  useListOwnerInstallationRequestsQuery,
  useCreateOwnerInstallationRequestMutation,
  useUpdateOwnerInstallationRequestMutation,
  useListTechInstallationRequestsQuery,
  useUpdateTechInstallationRequestMutation,
  useUpdateTechInstallationItemMutation,
  useApproveAllTechInstallationItemsMutation,
} = installationRequestsApi;

File: backend/app/schemas/installation_requests.py
"""Schemas for installation requests (owner ↔ technician workflow)."""
from datetime import datetime
from typing import Optional, Sequence
from uuid import UUID

from pydantic import BaseModel


class InstallationItemBase(BaseModel):
    """Base fields for an installation item."""

    room_id: Optional[UUID] = None
    coverage_type: str
    desired_device_count: int = 1
    notes: Optional[str] = None


class InstallationItemCreate(InstallationItemBase):
    """Create payload for an installation item."""

    pass


class InstallationItemResponse(InstallationItemBase):
    """Response schema for an installation item."""

    id: UUID
    proposed_device_type: Optional[str] = None
    status: str

    class Config:
        from_attributes = True


class InstallationRequestCreate(BaseModel):
    """Owner creates a new installation request."""

    home_id: Optional[UUID] = None
    notes: Optional[str] = None
    items: Sequence[InstallationItemCreate]


class InstallationRequestResponse(BaseModel):
    """Full installation request with nested items."""

    id: UUID
    home_id: UUID
    owner_id: UUID
    technician_id: Optional[UUID] = None
    status: str
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    items: list[InstallationItemResponse]

    class Config:
        from_attributes = True


class OwnerInstallationStatusUpdate(BaseModel):
    """Owner approves or requests changes."""

    action: str  # 'approve' or 'request_changes'
    notes: Optional[str] = None


class TechInstallationUpdate(BaseModel):
    """Technician updates request status or notes."""

    status: Optional[str] = None  # 'in_review','plan_ready','installed'
    notes: Optional[str] = None


class TechInstallationItemUpdate(BaseModel):
    """Technician updates a single room item within a request."""

    status: Optional[str] = None  # 'pending','approved','rejected','installed'
    desired_device_count: Optional[int] = None
    notes: Optional[str] = None


